### ISSUE TITLE
Configuration management can't know what modules are missing.

### ISSUE THREAD

**Issue by jenlampton** [backdrop/backdrop-issues#1811]
When you run an import containing configuration for a module that is disabled, you'll see something like the following error:

`The field "field_location" could not be created because it requires the "" module.`

Well, that's not very useful!  It's trying to tell us which module is needed, but it has no way of knowing.  Let's change that!

When saving a config file, we should include the machine name of the module that's responsible for the file.  Let's add it at the top along with the `_config_name`.  We could call it `_config_module`.

**Comment by klonos** [comment]
:+1:

**Comment by sentaidigital** [comment]
Can we add links or an "install all" button that adds the missing modules to the Project Browser install queue?

**Comment by jenlampton** [comment]
@sentaidigital yes, great idea!

**Comment by quicksketch** [comment]
> Can we add links or an "install all" button that adds the missing modules to the Project Browser install queue?

Now that would be some hot stuff! I love the idea of importing a whole site into a fresh environment, and having it automatically download and install the needed modules. That would be slick! But first, let's fix this part so that config files know which modules they need.

This also makes me wonder if the config files should have a list of dependencies, rather than a single module. e.g. a `_config_dependencies` entry rather than just `_config_module`. Then module extending config files (e.g. the `node.type.*` config files) could do something like:

```
$config = config('node.type.foo');
$config->set('my_module_setting', 'foo');
$config->addDependency('my_module');
$config->save();
```

Then importing the content type config file would know it needs not just the node module but also the extending module(s). Right now we don't actually track this at all, so we have no idea that a module that extends an existing config file would also be needed.

**Comment by jenlampton** [comment]
I've created a follow-up for the config file dependencies: https://github.com/backdrop/backdrop-issues/issues/4188

I'd love to focus on getting the module name into config files here, so we can update the UI, which right now seems a broken.

**Comment by alanmels** [comment]
If implemented then it would be easily possible (in core or special contrib module) to delete all the configuration files left by the modules which failed to clean up after they were uninstalled. I also like the idea. 

And maybe we should also consider to add a clean up cron job to core? Would be really nice to avoid messy configuration directories when some websites will start trying and uninstalling numerous modules.

**Comment by jenlampton** [comment]
> And maybe we should also consider to add a clean up cron job to core?

We would need to be careful that we aren't deleting information that's not intended to be deleted. I often will disable a module for a period of months when it's not needed (or, for testing purposes) but want to re-enable it later. I don't want it's config accidentally purged by a cleanup routine. Still a good idea! Can you please create a separate issue?

**Comment by alanmels** [comment]
> Can you please create a separate issue?

Here it is, Jen: https://github.com/backdrop/backdrop-issues/issues/4189 Please pay special attention to the fact that the configuration file for a properly uninstalled module would get deleted anyway. So we are talking about auto-deleting config files of only uninstalled modules, not of disabled ones.

**Comment by docwilmot** [comment]
I dont understand how to reproduce this.

**Comment by jenlampton** [comment]
@docwilmot I've added STR to the Original Post.

**Comment by klonos** [comment]
> Steps to reproduce:
> 
> 1. Install backdrop core on site A
> 2. Install Backdrop core + youtube module on site B
> 3. Configure youtube module on site B
> 4. Export config on site B
> 5. Import config in site A
> 6. Error about missing "unknown" module

I believe that #5163 will solve this issue here:
- After step 2 there will be a `system.extensions.json` file on site B, which will include all modules that have been enabled (including the youtube module).
- In step 4, `system.extensions.json` will be included with the rest of the exported config files
- In step 5, `system.extensions.json` will be cross-checked against the system db table for any modules that are enabled in site B but not in site A, and those will be enabled. -> no errors 🙂

**Comment by jenlampton** [comment]
I feel it's important that we solve this without being blocked by #5163 (which is a contentious issue). The current recommended solution is simple and wouldn't rely on anything else being in place.

**Comment by jenlampton** [comment]
Now that https://github.com/backdrop/backdrop-issues/issues/5163 is in and we've been using (and loving?) that for a while, I would like to return to this problem, because I still often run into issues with config files that are missing their modules. 

I have started on a PR at https://github.com/backdrop/backdrop/pull/5253, but it is incomplete. Pushing here for now because I do not know when I will get back to it, and did not want to loose my work. 

I am also open to feedback if anyone hates this approach :D
